crates/sui/src/upgrade_compatibility.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Error};
+use std::collections::BTreeSet;
+use thiserror::Error;
+
+use move_binary_format::{
+    compatibility::Compatibility,
+    compatibility_mode::CompatibilityMode,
+    file_format::AbilitySet,
+    normalized::{Enum, Function, Module, Struct},
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    language_storage::ModuleId,
+};
+use sui_json_rpc_types::{SuiObjectDataOptions, SuiRawData};
+use sui_protocol_config::ProtocolConfig;
+use sui_sdk::SuiClient;
+use sui_types::{base_types::ObjectID, execution_config_utils::to_binary_config};
+
+/// Check the upgrade compatibility of a new package with an existing on-chain package.
+pub async fn check_compatibility(
+    client: &SuiClient,
+    package_id: ObjectID,
+    compiled_modules: &[Vec<u8>],
+    protocol_config: ProtocolConfig,
+) -> Result<(), Error> {
+    let new_modules = compiled_modules
+        .iter()
+        .map(|b| CompiledModule::deserialize_with_config(b, &to_binary_config(&protocol_config)))
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|_| anyhow!("Unable to check compatibility of new module"))?;
```

Comment:
`anyhow` has a `context` function that allows you to wrap an error in an `anyhow::Error`, adding extra contextual information -- it might be helpful to you here:

```suggestion
        .context("Unable to to deserialize compiled module")?;
```